src/context.rs

```diff
@@ -0,0 +1,354 @@
+        // Add plan to queue and increment counter
+        let plan_id = self.plan_counter;
+        self.queue.push(TimedPlan {
+            time,
+            plan_id,
+            callback: Box::new(callback),
+        });
+        self.plan_counter += 1;
+        PlanId { id: plan_id }
+    }
+
+    pub fn cancel_plan(&mut self, id: PlanId) {
+        self.invalid_set.insert(id.id);
+    }
+
+    pub fn get_next_timed_plan(&mut self) -> Option<TimedPlan> {
+        loop {
+            let next_timed_plan = self.queue.pop();
+            match next_timed_plan {
+                Some(timed_plan) => {
+                    if self.invalid_set.contains(&timed_plan.plan_id) {
+                        self.invalid_set.remove(&timed_plan.plan_id);
+                    } else {
+                        return Some(timed_plan);
+                    }
+                }
+                None => {
+                    return None;
+                }
+            }
+        }
+    }
+}
+
+type Callback = dyn FnOnce(&mut Context);
+type EventHandler<E> = dyn Fn(&mut Context, E);
+pub struct Context {
+    plan_queue: PlanQueue,
+    callback_queue: VecDeque<Box<Callback>>,
+    event_handlers: HashMap<TypeId, Box<dyn Any>>,
+    immediate_event_handlers: HashMap<TypeId, Box<dyn Any>>,
+    plugin_data: HashMap<TypeId, Box<dyn Any>>,
+    time: f64,
+}
+
+impl Context {
+    pub fn new() -> Context {
+        Context {
+            plan_queue: PlanQueue::new(),
+            callback_queue: VecDeque::new(),
+            event_handlers: HashMap::new(),
+            immediate_event_handlers: HashMap::new(),
+            plugin_data: HashMap::new(),
+            time: 0.0,
+        }
+    }
+
+    pub fn add_plan(&mut self, time: f64, callback: impl FnOnce(&mut Context) + 'static) -> PlanId {
+        // TODO: Handle invalid times (past, NAN, etc)
+        self.plan_queue.add_plan(time, callback)
+    }
+
+    pub fn cancel_plan(&mut self, id: PlanId) {
+        self.plan_queue.cancel_plan(id);
+    }
+
+    pub fn queue_callback(&mut self, callback: impl FnOnce(&mut Context) + 'static) {
+        self.callback_queue.push_back(Box::new(callback));
+    }
+
+    fn add_plugin<T: Plugin>(&mut self) {
+        self.plugin_data
+            .insert(TypeId::of::<T>(), Box::new(T::get_data_container()));
+    }
+
+    pub fn get_data_container_mut<T: Plugin>(&mut self) -> &mut T::DataContainer {
+        let type_id = &TypeId::of::<T>();
+        if !self.plugin_data.contains_key(type_id) {
+            self.add_plugin::<T>();
+        }
+        let data_container = self
+            .plugin_data
+            .get_mut(type_id)
+            .unwrap()
+            .downcast_mut::<T::DataContainer>();
+        match data_container {
+            Some(x) => x,
+            None => panic!("Plugin data container of incorrect type"),
+        }
+    }
+
+    pub fn get_data_container<T: Plugin>(&self) -> Option<&T::DataContainer> {
```

Comment:
Unfortunately, no. The standard library has this same kind of pattern because &T and &mut T are very different things (the code just makes it look like they could be handled similarly).